vidarr-server/src/test/java/ca/on/oicr/gsi/vidarr/server/PriorityByWorkflowTest.java

```diff
@@ -0,0 +1,199 @@
+  }
+
+  @Test
+  public void testIfWorkflowRunWithHigherPriorityExists_thenWorkflowRunDoesNotLaunch() {
+    JsonNode higherPriority = mapper.valueToTree(4);
+    JsonNode lowerPriority = mapper.valueToTree(2);
+
+    sut.set(workflow, "qwerty", Optional.of(higherPriority));
+
+    Optional<String> requestError = sut.request(workflow, version, "abcdef",
+        Optional.of(lowerPriority)).apply(consumableResourceCheckerVisitor);
+
+    assertTrue(requestError.isPresent());
+    assertEquals(requestError.get(), "There are test workflows currently queued up with higher "
+        + "priority.");
+  }
+
+  @Test
+  public void testIfWorkflowRunWithLowerPriorityExists_thenWorkflowRunLaunches() {
+    JsonNode higherPriority = mapper.valueToTree(4);
+    JsonNode lowerPriority = mapper.valueToTree(2);
+
+    sut.set(workflow, "qwerty", Optional.of(lowerPriority));
+
+    Optional<String> requestError = sut.request(workflow, version, "abcdef",
+        Optional.of(higherPriority)).apply(consumableResourceCheckerVisitor);
+
+    assertTrue(requestError.isEmpty());
+  }
+
+  @Test
+  public void testIfWorkflowRunWithSamePriorityExists_thenWorkflowRunLaunches() {
+    JsonNode twoPriority = mapper.valueToTree(2);
+    JsonNode twoPriorityAlso = mapper.valueToTree(2);
+
+    sut.set(workflow, "qwerty", Optional.of(twoPriority));
+
+    Optional<String> requestError = sut.request(workflow, version, "abcdef",
+        Optional.of(twoPriorityAlso)).apply(consumableResourceCheckerVisitor);
+
+    assertTrue(requestError.isEmpty());
+  }
+
+  @Test
+  public void testIfWorkflowRunWIthEmptyPriorityIsSet_thenWorkflowRunIsAddedWithLowestPriority() {
+    JsonNode lowestPriority = mapper.valueToTree(1);
+
+    Optional<String> requestErrorForEmpty = sut.request(workflow, version, "abcdef",
+        Optional.empty()).apply(consumableResourceCheckerVisitor);
+
+    Optional<String> requestErrorForLowest = sut.request(workflow, version, "abcdef",
+        Optional.of(lowestPriority)).apply(consumableResourceCheckerVisitor);
+
+    assertTrue(requestErrorForEmpty.isEmpty());
+    assertTrue(requestErrorForLowest.isEmpty());
+
+  }
+
+  @Test
+  public void testIfResourceIsReleasedButNotLaunched_thenWorkflowRunIsAddedBackToWaiting() {
+
+    JsonNode higherPriority = mapper.valueToTree(4);
+    JsonNode lowerPriority = mapper.valueToTree(1);
+
+    Optional<String> requestErrorHigher = sut.request(workflow, version, "qwerty",
+        Optional.of(higherPriority)).apply(consumableResourceCheckerVisitor);
+
+    sut.release(workflow, version, "qwerty", Optional.of(higherPriority));
+
+    Optional<String> requestErrorLower = sut.request(workflow, version, "abcdef",
+        Optional.of(lowerPriority)).apply(consumableResourceCheckerVisitor);
+
+    assertTrue(requestErrorHigher.isEmpty());
+    assertTrue(requestErrorLower.isPresent());
+    assertEquals(requestErrorLower.get(), "There are test workflows currently queued up with higher "
```

Comment:
For this and other tests testing that a workflow run is added back to waiting, the code could change to returning an error _without_ adding back to the waiting list and this test would still pass. Can you get and inspect the WaitingState instead?

Reply:
If you don't want to make WaitingState public or use reflection to access it, then that's fine, but you're not actually testing "thenWorkflowRunIsAddedBackToWaiting"

Reply:
Yeah that's fair, I can clarify the wording. Checking for an error state is reflective of what ConsumableResourceChecker is doing so I'd like to keep these tests as is and preferably add on unit tests using reflect.